ledger-tool/src/main.rs

```diff
@@ -2074,13 +2112,48 @@ fn main() {
                 );
                 exit(1);
             }
-            println!("Purging data from slots {} to {}", start_slot, end_slot);
-            if no_compaction {
-                blockstore.purge_slots(start_slot, end_slot, PurgeType::Exact);
+            info!(
+                "Purging data from slots {} to {} ({} slots) (skip compaction: {}) (dead slot only: {})",
+                start_slot,
+                end_slot,
+                end_slot - start_slot,
+                no_compaction,
+                dead_slots_only,
+            );
+            let purge_from_blockstore = |start_slot, end_slot| {
+                blockstore.purge_from_next_slots(start_slot, end_slot);
+                if no_compaction {
+                    blockstore.purge_slots(start_slot, end_slot, PurgeType::Exact);
+                } else {
+                    blockstore.purge_and_compact_slots(start_slot, end_slot);
+                }
+            };
+            if !dead_slots_only {
+                let slots_iter = &(start_slot..=end_slot).chunks(batch_size);
+                for slots in slots_iter {
+                    let slots = slots.collect::<Vec<_>>();
+                    assert!(!slots.is_empty());
+
+                    let start_slot = *slots.first().unwrap();
+                    let end_slot = *slots.last().unwrap();
+                    info!(
+                        "Purging chunked slots from {} to {} ({} slots)",
+                        start_slot,
+                        end_slot,
+                        end_slot - start_slot
+                    );
+                    purge_from_blockstore(start_slot, end_slot);
+                }
             } else {
-                blockstore.purge_and_compact_slots(start_slot, end_slot);
+                let dead_slots_iter = blockstore
+                    .dead_slots_iterator(start_slot)
+                    .unwrap()
+                    .take_while(|s| *s <= end_slot);
+                for dead_slot in dead_slots_iter {
+                    info!("Purging dead slot {}", dead_slot);
+                    purge_from_blockstore(dead_slot, dead_slot);
+                }
             }
-            blockstore.purge_from_next_slots(start_slot, end_slot);
```

Comment:
this order was very broken, when the process is interrupted immediately after `purge_slots` and before `purge_from_next_slots`. It leaves super dangerous dangling references in meta....

Reply:
Well, when I was working around https://github.com/solana-labs/solana/pull/12350/files#diff-5e9f940ea065adfd3066ef0d8ef0cfb5029b5ba478d96fb484b36954e464505cR1609, I just skipped the fact check and corrected new code only.... and now that laziness bit me. ;)